### PR TITLE
feat(squash): add squash command and action

### PR DESCRIPTION
**Pull request by sileht**
Fixes MRGFY-284

Change-Id: I26d3c9bb5be07768cff182707309176880a18b6c

### CODE REVIEW

**Comment by sileht** [comment]
This pull request is part of a stack:
1. refactor: move some Git error logic to Gitter() ([#2833](https://github.com/Mergifyio/mergify-engine/pull/2833))
1. refactor: move UserTokensUser selection to UserTokens class ([#2834](https://github.com/Mergifyio/mergify-engine/pull/2834))
1. feat(squash): add squash command and action ([#2835](https://github.com/Mergifyio/mergify-engine/pull/2835))

File: mergify_engine/actions/squash.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+            )
+
+        if commit_title_and_message is not None:
+            title, message = commit_title_and_message
+            message = f"{title}\n\n{message}"
+        else:
+            message = f"{(await ctxt.pull_request.title)} (#{(await ctxt.pull_request.number)})"
+            message += "\n\n".join(
+                [commit["commit"]["message"] for commit in await ctxt.commits]
+            )
```
Comment (jd) [inline mergify_engine/actions/squash.py:83]: Pretty sure this is not how GitHub formats the squash message by default.

File: mergify_engine/squash_pull.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+    git = gitter.Gitter(ctxt.log)
+
+    try:
+        await git.init()
+        # NOTE(sileht): Bump the repository format. This ensures required
+        # extensions (promisor, partialclonefilter) are present in git cli and
+        # raise an error if not. Avoiding git cli to fallback to full clone
+        # behavior for us.
+        await git("config", "core.repositoryformatversion", "1")
```
Comment (jd) [inline mergify_engine/squash_pull.py:63]: wondering if we shouldn't do this in `git.init` at this point?

File: mergify_engine/squash_pull.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+    if ctxt.pull["commits"] <= 1:
+        return
+
+    try:
+        users = await user_tokens.UserTokens.select_users_for(ctxt, bot_account)
+    except user_tokens.UserTokensUserNotFound:
+        raise SquashFailure(
+            f"Unable to squash: user `{bot_account}` is unknown. "
+            f"Please make sure `{bot_account}` has logged in Mergify dashboard."
+        )
```
Comment (jd) [inline mergify_engine/squash_pull.py:139]: We have this elsewhere too, might be great to refactor at some point.

File: mergify_engine/squash_pull.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+            )
+        else:
+            return
+
+    ctxt.log.warning("unable to squash pull request: no tokens are valid")
+
+    if ctxt.pull_from_fork and ctxt.pull["base"]["repo"]["private"]:
+        raise SquashFailure(
+            "Squashing a branch for a forked private repository is not supported by GitHub"
+        )
```
Comment (jd) [inline mergify_engine/squash_pull.py:158]: Why is that so late in the process? We should do this first thing to bypass the retry loop?

**Review by jd — CHANGES_REQUESTED** [review]

File: mergify_engine/squash_pull.py (hunk 4 above)
Comment (sileht) [reply mergify_engine/squash_pull.py:158]: When we wrote this for branch_updater, we was not 100% confident about the conditions for this to work or not.
That why we wrote the message only after we try all tokens.

Comment (jd) [reply mergify_engine/squash_pull.py:158]: Can we check the log in prod to see if that happen with the branch updater?

**Review by jd — APPROVED** [review]

**Comment by jd** [comment]
@mergifyio update

File: mergify_engine/squash_pull.py (hunk 4 above)
Comment (sileht) [reply mergify_engine/squash_pull.py:158]: I don't think we can catch that in logs.

**Comment by jd** [comment]
@mergifyio update

**Pull request merged**